tests/js/searchtools.js

```diff
@@ -85,6 +102,52 @@ describe('Basic html theme search', function() {
 
   });
 
+  describe('search result ranking', function() {
+
+    /*
+     * These tests should not proscribe precise expected ordering of search
+     * results; instead each test case should describe a single relevance rule
+     * that helps users to locate relevant information efficiently.
+     *
+     * If you think that one of the rules seems to be poorly-defined or is
+     * limiting the potential for search algorithm improvements, please check
+     * for existing discussion/bugreports related to it on GitHub[1] before
+     * creating one yourself. Suggestions for possible improvements are also
+     * welcome.
+     *
+     * [1] - https://github.com/sphinx-doc/sphinx.git/
+     */
+
+    it('should score an object-name match above a page-title match', function() {
+      eval(loadFixture("titles/searchindex.js"));
+
+      expectedRanking = [
+        ['index', 'relevance', '#module-relevance'],  /* py:module documentation */
+        ['relevance', 'Relevance', ''],  /* main title */
+      ];
```

Comment:
I'm questioning my own test case a bit here: is it really true that an object name should match about a main document title in the general case?

Reply:
Questioning this further: is the `other.Example.relevance` attribute more relevant to the search query `relevance` than the document with title `Relevance`?  I would probably suggest not (and perhaps I should have chosen a different word to use for these test cases).

Reply:
Yep, after some more thought I think I'd suggest these rules -- all in the context of exact-match names (partial matches are worth considering, but later):

  * A ['high priority' object match](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/themes/basic/static/searchtools.js#L34) (such as a Python module name) should rank above a main document title match.
  * A main document title match should rank above a normal priority object match (such as a Python class attribute).

Achieving this does require one more change to the search code; I've prototyped that locally to produce the results displayed below:

![image](https://github.com/sphinx-doc/sphinx/assets/55152140/a741f8a1-a498-4eda-89d5-d1a0d8fcbe1e)

Reply:
> Achieving this does require one more change to the search code; I've prototyped that locally to produce the results displayed below:

My mistake there; no code change is required.  I've updated the test suite in this branch to include the suggested rules, though.